File: tools/glosh-remote-spike/app/src/main/java/com/glosh/remote/spike/guide/autopilot/AutopilotUiModel.java
package com.glosh.remote.spike.guide.autopilot;

import com.glosh.remote.spike.guide.accessibility.NodeSnapshot;

import java.util.Collections;
import java.util.EnumMap;
import java.util.Map;

public final class AutopilotUiModel {
    public enum TargetKey {
        SOFTWARE_INFO("software_info"),
        BUILD_NUMBER("build_number"),
        WIRELESS_DEBUGGING("wireless_debugging"),
        WIRELESS_DEBUGGING_TOGGLE("wireless_debugging_toggle"),
        PAIR_WITH_CODE("pair_with_code"),
        NETWORK_CONFIRM_POSITIVE("network_confirm_positive");

        private final String plannerKey;

        TargetKey(String plannerKey) {
            this.plannerKey = plannerKey;
        }

        public String plannerKey() {
            return plannerKey;
        }
    }

    public record MatchedTarget(
            TargetKey key,
            NodeSnapshot node,
            AutopilotContract.Confidence confidence,
            int score,
            Integer runnerUpScore,
            boolean clickable,
            boolean unique,
            boolean marginOk) {
    }

    public record ClassifiedScreen(
            AutopilotContract.Screen screen,
            AutopilotContract.Confidence confidence,
            Map<TargetKey, MatchedTarget> targets,
            Boolean wirelessEnabled,
            boolean policyBlocked) {
        public ClassifiedScreen {
            EnumMap<TargetKey, MatchedTarget> copy = new EnumMap<>(TargetKey.class);
            if (targets != null) {
                copy.putAll(targets);
            }
            targets = Collections.unmodifiableMap(copy);
        }

        public MatchedTarget target(TargetKey key) {
            return targets.get(key);
        }
    }

    private AutopilotUiModel() {
    }
}

File: docs/artifacts/remote-install-autopilot/reference/AutopilotUiModel.kt
package reference.autopilot

data class IntRect(val left:Int,val top:Int,val right:Int,val bottom:Int) {
    val area: Long get() = (right-left).coerceAtLeast(0).toLong() * (bottom-top).coerceAtLeast(0).toLong()
}

enum class WindowKind { APPLICATION, ACCESSIBILITY_OVERLAY, INPUT_METHOD, OTHER }

data class UiNodeSnapshot(
    val path: String,
    val text: String? = null,
    val contentDescription: String? = null,
    val viewId: String? = null,
    val className: String? = null,
    val clickable: Boolean = false,
    val checkable: Boolean = false,
    val checked: Boolean? = null,
    val visible: Boolean = true,
    val enabled: Boolean = true,
    val bounds: IntRect = IntRect(0,0,0,0),
    val ancestorTexts: List<String> = emptyList(),
)

data class UiWindowSnapshot(
    val id: Int,
    val packageName: String?,
    val kind: WindowKind,
    val focused: Boolean,
    val nodes: List<UiNodeSnapshot>,
)

data class TrustedSettingsSnapshot(
    val windowId: Int,
    val packageName: String,
    val nodes: List<UiNodeSnapshot>,
    val fingerprint: String,
)

sealed interface WindowSelection {
    data class Selected(val snapshot: TrustedSettingsSnapshot): WindowSelection
    data class Rejected(val reason: String): WindowSelection
}

enum class TargetKey {
    SOFTWARE_INFO,
    BUILD_NUMBER,
    WIRELESS_DEBUGGING,
    WIRELESS_DEBUGGING_TOGGLE,
    PAIR_WITH_CODE,
    NETWORK_CONFIRM_POSITIVE,
}

data class MatchedTarget(
    val key: TargetKey,
    val nodePath: String,
    val confidence: Confidence,
    val score: Int,
    val runnerUpScore: Int?,
    val clickable: Boolean,
    val unique: Boolean,
    val marginOk: Boolean,
)

data class ClassifiedScreen(
    val screen: Screen,
    val confidence: Confidence,
    val targets: Map<TargetKey, MatchedTarget>,
)
